datafusion/core/src/datasource/file_format/write.rs

```diff
@@ -315,58 +300,237 @@ pub(crate) async fn create_writer(
     }
 }
 
+/// Serializes a single data stream in parallel and writes to an ObjectStore
+/// concurrently. Data order is preserved. In the event of an error,
+/// the ObjectStore writer is returned to the caller in addition to an error,
+/// so that the caller may handle aborting failed writes.
+async fn serialize_rb_stream_to_object_store(
```

Comment:
I have some ideas how to simplify this code, which I will try out shortly, but I also think it can be merged like this too.

Reply:
I tried to rewrite this as a `futures::stream` computation using `buffered` -- but I got stuck on some "higher-ranked lifetime error" so I think this is about as good as it is going to get